reactioncommerce/reaction
# Additional navigation visibility options

Resolves #5146  
Impact: **minor**  
Type: **feature**

## Issue

Navigation items need some additional visibility options

## Solution

Add the following options
- `Show in storefront (isVisibile)` - show or hide from API / storefront etc
- `Admin access only (isPrivate)` - only show to admins in storefront, if `isVisibile: true`
- `Secondary nav only (isSecondary)` - show in the secondary nav, useful if you use the same nav for the main / sidebar, but want the sidebar to have some additional items.

## Breaking changes

- `navigationTreeById` now accepts an object containing params `{ language, navigationTreeId, shouldInlcudeSecondary }` instead of individual params.
- `draftItems` is admin only now so to not expose work-in-progress changes to the public

## Testing
1. Add navigations items to the navigation tree in the `Navigation` editor
1. Edit an item in the tree
1. Set `Show in storefront`, `Admin access only`, `Secondary nav only` options for various items
1. Be sure to "Save Changes" in the toolbar at the top to ensure all items get published
1. Ensure the following queries give you the items you expect

Keep in mind a few things for these queries.
- `items` gets filtered by `isVisible`, `isPrivate` automatically and by the `shouldShowSecondary` arg
- `draftItems` is admin only now and is not filtered by permissions

```graphql
{
  primaryShop {
    defaultNavigationTree(language: "en", shouldIncludeSecondary: true) {
      name
      items {
        isVisible
        isPrivate
        isSecondary
        navigationItem {
          data {
            content {
              value
            }
          }
        }
      }
      draftItems {
        isVisible
        isPrivate
        isSecondary
      }
    }
  }
}
```
---

Get the default navigation tree id
```graphql
{
  primaryShop {
    defaultNavigationTreeId
  }
```

Use the ID for this query
```graphql
{
  navigationTreeById(
    id: "<nav-tree-id>,
    language: "en",
    shouldIncludeSecondary: true
  ) {
    items {
        isVisible
        isPrivate
    	isSecondary
         navigationItem {
          data {
            content {
              value
            }
          }
        }
      items {
        isVisible
        isPrivate
        isSecondary
         navigationItem {
          data {
            content {
              value
            }
          }
        }
      }
    }
  }
}
```
